File: python/day7.py
import re
import sys
from functools import partial
from itertools import chain

hypernets = re.compile(r'\[([^]]*?)\]').findall
supernets = re.compile(r'\[[^]]*?\]').split


def pieces_of(n, s):
    return [s[i:i + n] for i in xrange(len(s) - n + 1)]


def mapcat(f, xs):
    return list(chain.from_iterable(map(f, xs)))


def is_palindrome(s):
    return s[0] == s[-1] and s[0] != s[1]


def is_abba(s):
    sn = mapcat(partial(pieces_of, 4), supernets(s))
    hn = mapcat(partial(pieces_of, 4), hypernets(s))
    return filter(is_palindrome, sn) and not filter(is_palindrome, hn)


def solution1(data):
    return len(filter(is_abba, data))


def is_aba_bab(s):
    sn = mapcat(partial(pieces_of, 3), supernets(s))
    hn = mapcat(partial(pieces_of, 3), hypernets(s))
    for aba in filter(is_palindrome, sn):
        bab = aba[1] + aba[0] + aba[1]
        if bab in hn:
            return True


def solution2(data):
    return len(filter(is_aba_bab, data))


examples1 = ["abba[mnop]qrst",
             "abcd[bddb]xyyx",
             "aaaa[qwer]tyui",
             "ioxxoj[asdfgh]zxcvbn"]

examples2 = ["aba[bab]xyz",
             "xyx[xyx]xyx",
             "aaa[kek]eke",
             "zazbz[bzb]cdb"]

if __name__ == '__main__':
    puzzle = list(sys.stdin)
    print solution1(puzzle)
    print solution2(puzzle)

File: clojure/src/advent2016/day7.clj
(ns advent2016.day7
  (:require [clojure.string :as s]
            [advent.core :refer [->lines]]
            [advent2016.core :refer [read-puzzle]]))

(def example1 ["abba[mnop]qrst"
               "abcd[bddb]xyyx"
               "aaaa[qwer]tyui"
               "ioxxoj[asdfgh]zxcvbn"])
(def example2 ["aba[bab]xyz"
               "xyx[xyx]xyx"
               "aaa[kek]eke"
               "zazbz[bzb]cdb"])
(def puzzle (read-puzzle "day7" ->lines))

(defn hypernet-seqs [s] (mapcat rest (re-seq #"\[([^]]*?)\]" s)))
(defn supernet-seqs [s] (s/split (s/replace s #"\[[^]]*?\]" "|") #"\|"))


(defn abba? [s]
  (re-seq #"(?:([a-z])(?!\1)([a-z])\2\1)+" s))
(defn abba [s]
  (if (every? nil? (map abba? (hypernet-seqs s)))
    (some identity (map abba? (supernet-seqs s)))))

(def solution1 (->> puzzle
                    (keep abba)
                    count))


(defn aba? [[a b c]]
  (and (= a c)
       (not= a b)))

(defn one-aba-bab? [hns sn]
  (->> (partition 3 1 sn)
       (filter aba?)
       (filter (fn [[a b]]
                 (some #(s/includes? % (str b a b)) hns)))))

(defn aba-bab? [s]
  (mapcat (partial one-aba-bab? (hypernet-seqs s))
          (supernet-seqs s)))

(def solution2
  (->> puzzle
       (map aba-bab?)
       (filter (comp not empty?))
       count))
